### ISSUE TITLE
KeePassXC does not exit on reboot/shutdown

### ISSUE THREAD

**Issue by yennor** [keepassxreboot/keepassxc#4605]
## Expected Behavior
When rebooting or shuting down the computer, keepassxc should quit properly.

## Current Behavior
Very often keepassxc doesn't quit. (probably ignoring the kill signal? or whatever is send).
So I get the nice message of Systemd waiting for a stop job, for 90 seconds. Until keepassxc gets killed. In the log it looks like that:

> Apr 15 10:51:43 RMMbook systemd[1]: session-3.scope: Stopping timed out. Killing.
> Apr 15 10:51:43 RMMbook systemd[1]: session-3.scope: Killing process 1496 (keepassxc) with signal SIGKILL.
> Apr 15 10:51:43 RMMbook systemd[1]: session-3.scope: Failed with result 'timeout'.
> 

Keepassxc does that since at least version 2.3.x

## Debug Info
KeePassXC - Version 2.5.4
Revision: dcca5aa

Qt 5.12.5
Debugging mode is disabled.

Operating system: Debian GNU/Linux 10 (buster)
CPU architecture: x86_64
Kernel: linux 5.4.0-4-amd64

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- Secret Service Integration

Cryptographic libraries:
 libgcrypt 1.8.4

**Comment by droidmonkey** [comment]
We added several fixes to the shutdown sequence in 2.5.x. I personally tested these on Gnome, KDE, and XFCE desktop envs. Can you provide some additional context on your application settings (screenshot is ok) and desktop env?

**Comment by yennor** [comment]
Desktop env:
Kde Plasma :5.17.5-4
Debian stable with too many mixins from testing (eg. kde plasma). But about 4 weeks ago I've still had a pure Debian stable system and it showed the same behaviour but then I urgently needed same newer packages...
I've just added all the general settings
![settings1b](https://user-images.githubusercontent.com/6601243/79508516-52c12880-7fff-11ea-9762-92802791c7f9.png)
![settings 2](https://user-images.githubusercontent.com/6601243/79508522-55238280-7fff-11ea-91a4-01c89f583302.png)
![settings 1](https://user-images.githubusercontent.com/6601243/79508527-55bc1900-7fff-11ea-9212-0f91fbc16f95.png)

I don't think it should make a difference in that context, but I'm using the ubuntu disco package from https://launchpad.net/~phoerious/+archive/ubuntu/keepassxc

If you need anymore information just ask...

**Comment by yennor** [comment]
I've forgot to add, that it doesn't happen always. But most of the time (>80% ?  I don't restart my laptop that often, mostly I just send it to sleep, so I can't give a reliable estimate)...

**Comment by droidmonkey** [comment]
If you can live without "Minimize instead of app exit" that will resolve your issue

**Comment by yennor** [comment]
Not really. I usually close those applications by using alt-f4.
And other applications (music player, download manager) which minimize on close to the system tray don't have the same problem. so it should be possible to implement it.

**Comment by droidmonkey** [comment]
As I said before, we did add a fix for this issue and for whatever reason it is not working for you.

**Comment by Riamse** [comment]
I'm affected too. "Minimise instead of app exit" unchecked. KDE Plasma 5.18.4, btw I use Arch

**Comment by yan12125** [comment]
For me shutdown issues happen less often after https://github.com/keepassxreboot/keepassxc/pull/4309, but I was hit by it again today. I guess that PR does not fix all shutdown issues.

Environment: LXQt 0.14.1 + Qt 5.14.2 on Arch Linux. KeePassXC (2.6.0-Snapshot revision 5795051) is built from [the script on AUR](https://aur.archlinux.org/packages/keepassxc-git).

**Comment by ctsde-markus** [comment]
Hi there!

On macOS 10.13 and 10.14 i am hit with the same problem. When KeepassXC is still running when i use Apple -> Logout i get a message that the logout was interrupted by KeepassXC. So it's not a linux only issue, i guess.

**Comment by CocaineJohnsson** [comment]
KeePassXC 2.5.4, kernel 5.6.6 on Arch running AwesomeWM.

On system shutdown, do we receive a WM_DELETE_WINDOW or a SIGTERM? I'd suspect we'd receive a SIGTERM since init doesn't really know or care what is and isn't an X-client.

Notably on virtually any *nix platform you can manually override the SIGTERM handler as such (C, also legal C++) `signal(SIGTERM, signal_handler); //signal_handler is void (*)(int signum)`, if using something like SDL2 it will by default trap a few different signals as well as WM_DELETE_WINDOW in the SDL_QUIT event. This is sometimes desirable but for correct shutdown it's worthwhile handling SIGTERM specially, maybe SIGABRT too (though in my experience that shouldn't be necessary unless dealing with poorly coded abort-zealous libraries)

You might also be interested in SIGINT, i.e keyboard interrupt (typically CTRL+C from controlling terminal), these are generally the ones most likely to be used to terminate the application.

Though init's shutdown job will almost certainly use SIGTERM followed by waiting for some timeout (probably 120s, at least on most systemd based environments) and then forcefully ending you with SIGKILL.

Probably in this case it would be worthwhile to trap SIGTERM, if there is unsaved data either prompt the user to save via gui or just flush to disk for them (or assume they didn't want to save, whatever floats your boat).

The signature for a signal handler function is `void signal_handler(int signum)`, you can make one that doesn't take parameters but that is UB iirc.

At least that's what I'd infer, relevant journalctl output:

> Apr 27 06:09:08 Valhalla systemd[4721]: pulseaudio.service: Succeeded.
> Apr 27 06:09:12 Valhalla systemd[1]: lightdm.service: Succeeded.
> Apr 27 06:09:12 Valhalla systemd[1]: Stopped Light Display Manager.
> Apr 27 06:09:12 Valhalla systemd[1]: Stopping Getty on tty1...
> Apr 27 06:09:12 Valhalla systemd[1]: getty@tty1.service: Succeeded.
> Apr 27 06:09:12 Valhalla systemd[1]: Stopped Getty on tty1.
> Apr 27 06:09:12 Valhalla systemd[1]: Removed slice system-getty.slice.
> Apr 27 06:12:08 Valhalla systemd[1]: session-2.scope: Stopping timed out. Killing.
> Apr 27 06:12:08 Valhalla systemd[1]: session-2.scope: Killing process 4843 (keepassxc) with signal SIGKILL.
> Apr 27 06:12:08 Valhalla systemd[1]: session-2.scope: Killing process 5104 (QDBusConnection) with signal SIGKILL.
> Apr 27 06:12:08 Valhalla systemd[1]: session-2.scope: Killing process 5111 (gmain) with signal SIGKILL.
> Apr 27 06:12:08 Valhalla systemd[1]: session-2.scope: Killing process 5113 (gdbus) with signal SIGKILL.
> Apr 27 06:12:08 Valhalla systemd[1]: session-2.scope: Failed with result 'timeout'.
> Apr 27 06:12:08 Valhalla systemd[1]: Stopped Session 2 of user cocaine_johnsson

As a slight aside, you almost certainly know this already but I'd rather mention it just in case than omit important information: whatever signal handler was defined *last* overwrites the previous one, so if some library sets a handler for SIGTERM (e.g SDL2) you can override that by invoking signal() yourself for the appropriate signal(s)

**Comment by droidmonkey** [comment]
@CocaineJohnsson what you described is 100% already in the code base! 
https://github.com/keepassxreboot/keepassxc/blob/dcca5aa0f0ee2f2b8685787c01f934cb2b506ef7/src/gui/Application.cpp#L159-L208

Second, testing whether KeePassXC properly exits when a SIGTERM or SIGKILL message is sent to it is trivial. Just run KeePassXC, then open a terminal and issue:
```
# Send SIGTERM to KeePassXC process
kill -15 `pgrep keepassxc`

# Send SIGKILL to KeePassXC process
kill -9 `pgrep keepassxc`
```

I just tested sending these signals to KeePassXC with minimize instead of app exit and it quit every single time without hesitation.

**Comment by droidmonkey** [comment]
Either way, on Linux the actual path for shutdown is typically through XCB session management. Qt handles this interaction and we receive a `closeEvent` through our MainWindow. We know that this event is distinct from user interaction because of this line of code:

https://github.com/keepassxreboot/keepassxc/blob/dcca5aa0f0ee2f2b8685787c01f934cb2b506ef7/src/gui/MainWindow.cpp#L1003

specifically `qApp->isSavingSession()`. Again, reasons for failing to close include not saving your database prior to reboot.

**Comment by CocaineJohnsson** [comment]
I see, I see. My bad... then I'm somewhat at a loss for the obvious cause, shutdown should issue a SIGTERM and that handling looks correct at a glance -- though it's still a bit odd that it would then stall the shutdown job (inconsistently, mind), mildly annoying is that it treats CTRL+Q the same way as a window close event, though I'm not super familiar with the windowing code specifically so it's quite possible that it generates a WM_DELETE_WINDOW instead of SIGTERM or SIGINT

My bad, sorry for wasting your time... though that at least implies that for *whatever* reason the shutdown job sends a WM_DELETE_WINDOW (or similar\*) instead of a SIGTERM

\* Curiously how does it handle XDestroyWindow (or any DestroyNotify event)? I doubt the DM would send that to a child on death but... nothing else comes to mind, `XKillClient` is handled correctly as tested by xkill -id $XID

Either way, stalling a shutdown job is mildly problematic -- though for now I'll just use the workaround.

> Again, reasons for failing to close include not saving your database prior to reboot.

I strongly disagree. It is no doubt probably correct for some configurations, but with my configuration in particular it is not. It is set to autosave on any change, there is simply no possible case in which there are pending unsaved changes to the database that aren't committed.
Even *if* it were saving at the time it takes a couple of seconds to save, this shouldn't stall noticeably.

`if (config()->get("GUI/MinimizeOnClose").toBool() && !m_appExitCalled && !isHidden() && !qApp->isSavingSession())`

Assumption: `isHidden` is a boolean returning true if we are minimized.

Observation: shutdown was still stalled, with no pending changes and with CTRL+Q issued to KeePassXC approximately 4 seconds prior to `$ shutdown now`

The database remains unchanged since Apr 04, so there should be no pending or processing write to speak of, effectively eliminating it as a factor here.

Therefore I will infer that isSavingSession must return false. This points to the issue lying solely with the minimize on close setting, does it not (more specifically, stalling shutdown on pending writes is more than reasonable since we'd likely want to issue a notice to the user to save their data if at all possible)

I'm unfamiliar with Qt in general, is there any way to distinguish between what generated a close event? I'd presume there must be as other softwares with a similar option appear to not stall shutdown (in my experience, anyway -- notably many instant messaging clients and voip clients have options like this)

Given that the stalling is intermittent, is this possibly a bug in LightDm? Qt? systemd? Does this happen on SysV based installations? upstart? initng? busybox? others? (if it doesn't that still doesn't indicate systemd per se, rather it would imply an issue with the shutdown script which is likely provided by the login manager and/or display manager)

I can't really see any obvious problem with any of the overridden Qt methods, nothing that should obviously stall a shutdown job anyway (given the assumption of a sane system), the intermittency doesn't help with figuring it out either.

[Systemd issue #1615](https://github.com/systemd/systemd/issues/1615) (March/Apr 2016) implies it may be a broader issue with Qt and/or systemd given that plasma may also stall it. A comment from 2017 has this notable snippet

> and X stops before kded5 which somehow leads to latter to ignore SIGTERM.

This appears to match my logs fairly well, the display manager shuts down, and KeePassXC decides to not die.

The relevant log from that comment

> May 30 00:35:05 wahnsinn systemd[1]: Stopped Manage swap spaces on zram, files and partitions..
> May 30 00:35:07 wahnsinn sddm[687]: Display server stopped.
> May 30 00:35:07 wahnsinn sddm[687]: Running display stop script  "/usr/share/sddm/scripts/Xstop"
> May 30 00:35:07 wahnsinn sddm[687]: QProcess: Destroyed while process ("/usr/lib/sddm/sddm-helper") is still running.
> May 30 00:35:07 wahnsinn systemd[1]: Stopped Simple Desktop Display Manager.
> May 30 00:36:32 wahnsinn systemd[1]: session-c2.scope: Stopping timed out. Killing.
> May 30 00:36:32 wahnsinn systemd[1]: session-c2.scope: Killing process 1373 (kded5) with signal SIGKILL.
> May 30 00:36:32 wahnsinn systemd[1]: Stopped Session c2 of user gross.
> May 30 00:36:32 wahnsinn systemd[1]: session-c2.scope: Unit entered failed state.

I don't think it's likely to be a Systemd bug, more likely a lightdm and/or keepassxc bug. Possibly a Qt bug though Qt is widely used so that seems potentially unlikely.

It would be interesting to know what display managers the other reporters are using, if any.

Odd that the process isn't immediately reaped when its parent dies though, normally you'd expect the x-server going poof to reap the process immediately on merit of the x-server emitting a SIGTERM to all processes running under it.

As far as I care this implies a bug in the signal handling...

Does quitSignalReceived(); emit m_appExitCalled or something with functionally identical effect (i.e anything that takes it into the same code path as checking if MinimizeOnClose)? Or does this go straight to cleaning up resources and shutting down the application?

**Comment by droidmonkey** [comment]
I just tested XFCE with LightDM and KeePassXC (2.6.0-develop) closed without issue. KeePassXC 2.5.4 did NOT close properly.

**Comment by happnatious1** [comment]
I randomly see this issue on Linux mint 19.3 cinnamon. keepassxc 2.5.4

**Comment by droidmonkey** [comment]
I think I know what is happening, this is not an issue with the FileWatcher but an issue with ASyncTask itself. We should be listening to the application quit message to break the event loop because when the app quits in the middle of an asynctask, it never receives the response from the spawned thread that is forcibly killed.

**Comment by droidmonkey** [comment]
Ehhh ran some debug scenarios and am having a hard time pinning down the real issue.

**Comment by FuturePilot** [comment]
I'm seeing the same thing with 2.5.4 on Kubuntu 20.04. When I try to quit the program from Database -> Quit, the window disappears, but the notification area icon is still there and the process is still there. Clicking the icon in the notification area does nothing. Giving the process a SIGTERM doesn't quit the process. I have to give it a SIGKILL to get it to quit.

**Comment by veehexx** [comment]
seems like this could be distro specific. i've been subscribed to this report for a few months, and over the last month (since 2.5.4?) running Fedora 32 its been reliable.
on F31 and pre-2.5.4 (assuming 2.5.4 was the fixed version for me) i could gurantee keepassxc would not end nicely and cause long reboot/shutdown times, but ONLY after the laptop had been running for a while. i didnt get to test enough before it being resolved for me, but it was either after the laptop had been running for an hour or 2, or through sleep/hibernate where uptime was long.
If i clean booted, started XC (& login to db), and then immediately reboot/shutdown i'd never see the issue.

from what i remember closing keepassxc before shutdown/reboot was always graceful and thats as far as i got before updates came down.

**Comment by konomikitten** [comment]
This bug remains the bane of my existence I almost always have to issue `pkill -9 keepassxc` everytime I reboot my PC to avoid a 120 second delay as systemd kills off the keepassxc process.

```
KeePassXC - Version 2.6.1
Revision: 9a35bba

Qt 5.14.2
Debugging mode is disabled.

Operating system: Debian GNU/Linux bullseye/sid
CPU architecture: x86_64
Kernel: linux 5.7.0-3-amd64

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (only unsigned sharing)
- YubiKey
- Secret Service Integration

Cryptographic libraries:
 libgcrypt 1.8.6
```

**Comment by droidmonkey** [comment]
Or you could just quit keepassxc before shutting down.

**Comment by konomikitten** [comment]
@droidmonkey I do close it every time I shut down/reboot it still has a hanging pid with no GUI showing. The only way to get rid of it is with sigkill.

**Comment by droidmonkey** [comment]
Most likely fixed in 2.6.2, I found a potential issue with how we were allowing the application to still remain running even though all windows were closed and no Taskbar icon.

**Comment by konomikitten** [comment]
@droidmonkey hopefully Debian will get around to packaging it then thanks for the information.

**Comment by tommrwatts** [comment]
I appear to have a related issue on Windows 10 where my SSH key will somehow remain loaded in the [Windows variant of OpenSSH](https://github.com/PowerShell/Win32-OpenSSH/) across reboots, unless I explicitly lock or close the database before rebooting.

I noticed a couple of times that I could suddenly push and pull without unlocking my database after starting Windows. When I explicitly open and then lock my database again, this is no longer possible; it is as if my key got loaded into OpenSSH the previous day and stays in there until the next - which perhaps is weird in itself, I wouldn't expect the SSH agent to maintain the key, even if KeePassXC did not remove it.

**Comment by yan12125** [comment]
> I wouldn't expect the SSH agent to maintain the key, even if KeePassXC did not remove it.

That is how Microsoft's ssh-agent works. Some rationale can be found at https://github.com/PowerShell/Win32-OpenSSH/issues/1056#issuecomment-362494167

**Comment by littlebunyip** [comment]
Having this same issue. Run keepassxc minimised to system tray all the time. Power off machine without manually "quiting" keepassxc.

KeePassXC - Version 2.5.4
Revision: dcca5aa

Qt 5.11.3
Debugging mode is disabled.

Operating system: BunsenLabs GNU/Linux 10.5 (Lithium)
CPU architecture: x86_64
Kernel: linux 4.19.0-14-amd64

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (only unsigned sharing)
- YubiKey
- Secret Service Integration

Cryptographic libraries:
 libgcrypt 1.8.4

**Comment by droidmonkey** [comment]
You are out of date

**Comment by littlebunyip** [comment]
The latest debian package on the keepassxc download page points to debian sid (unstable). I am not running on the 'unstable' branch.

Version for buster (stable) is as the following from the output of:

`sudo apt-cache show keepassxc`

Package: keepassxc
Version: 2.5.4+dfsg.1-1~bpo10+1

snip ... and...

Package: keepassxc
Version: 2.3.4+dfsg.1-1

**Comment by droidmonkey** [comment]
No doubt, but this bug is fixed in later versions and you aren't running it. Use the appimage, snap, or ppa to get to the latest version.

**Comment by jschrod** [comment]
@droidmonkey In which version is this bug fixed?

I am running keepassxc 2.6.4, compiled from source at February 3, 2021, i.e., 2 weeks ago.
If I close the programm, the process keeps running and can only be killed with SIGHUP.

See my comments from June last year above; the situation hasn't changed.

**Comment by littlebunyip** [comment]
Thanks, i'll pass on those options. I only posted as I thought it was still open as per the github issue status.

**Comment by droidmonkey** [comment]
@jschrod if you are explicitly closing keepassxc and it's not closing then you have a unique problem. Are you sure it is not closing to system tray?

**Comment by jschrod** [comment]
@droidmonkey In the 1st version of your comment you asked, if I'm sure that it is not closing to system tray.

I am sure, since I don't have a system tray. I don't use KDE, GNOME, or Xfce -- I have an X server with a plain simple window manager (fvwm), being literally an old Unix grey-beard. I start keepassxc from the command line, and I use the top-level menu entry "Database / Quit" to close it. Is this action supposed to just stop the program, or does it more?

The problem appears only after I've been running the program for a few days. I just stopped it for my previous comment to see if the problem still happens; it does. I will come back with traces and gdb info where it hangs.

**Comment by droidmonkey** [comment]
Thank you, yah I didn't see your original comment from before because github hit it in the "XX Hidden Comments" sorry about that

**Comment by droidmonkey** [comment]
This should no longer be an issue, we fixed a couple shutdown issues in develop.

**Comment by saddy001** [comment]
Is this merged into release 2.6.6 already? Because I have the same issue.

Edit: I see the release was at the day this issue has been closed. So propably it's not merged.

**Comment by kzar79** [comment]
I am having this same problem with release 2.7.6

Every time I shutdown/restart the system I have to wait 2 minutes. If I first close KeePassXC it is immediate.

I also discovered another interesting thing: if I shutdown the system while the screen is locked, I do not have to wait. 

I know the bug is closed but it is still happening.

**Comment by JechoJekov** [comment]
It is still happening in 2.7.10. I am using it on openSUSE Tumbleweed and Leap 16. It happens even if the database is locked.

**Comment by cassandra-aq** [comment]
Still happening in 2.7.12 🤷
